morph_tool/morphdb.py

```diff
@@ -0,0 +1,346 @@
+'''Provides two classes to work with neurondb files
+The main use is to provide a neuronDB loader and a method to retrieve
+information as a dataframe.
+
+Example:
+
+old_release = MorphDB.from_neurondb(Path("/realease/one/neuronDB.xml"), label='old-release')
+new_release = MorphDB.from_neurondb(Path("/realease/two/neuronDB.xml"), label='new-release')
+total = old_release + new_release
+
+# neurondb information
+print(total.df)
+
+# combined information (neurondb + morph_stats)
+print(total.features({'neurite': {'section_lengths': ['max']}}))
+'''
+import logging
+from pathlib import Path
+from typing import Any, Dict, List, Optional, Union
+
+import pandas as pd
+import xmltodict
+from neurom.apps.morph_stats import extract_dataframe
+from toolz import isiterable
+
+from morph_tool.utils import iter_morphology_files
+
+L = logging.getLogger(__name__)
+
+
+# Attributes found in the <repair></repair> block of the XML
+BOOLEAN_REPAIR_ATTRS = [
+    'use_axon',
+    'use_dendrites',
+    'axon_repair',
+    'dendrite_repair',
+    'basal_dendrite_repair',
+    'tuft_dendrite_repair',
+    'oblique_dendrite_repair',
+    'unravel',
+    'use_for_stats',
+]
+
+COLUMNS = ['name', 'mtype', 'msubtype', 'mtype_no_subtype',
+           'layer', 'label', 'path'] + BOOLEAN_REPAIR_ATTRS + ['axon_inputs']
+
+
+def find_morph(folder: Path, stem: str) -> Optional[Path]:
+    '''Returns the path to a morphology in morphology_dir matching stem.
+
+    If no morphology is found, returns None
+    '''
+    for ext in {'.asc', '.ASC', '.h5', '.H5', '.swc', '.SWC'}:
+        path = folder / (stem + ext)
+        if path.exists():
+            return path
+    return None
+
+
+class MorphInfo:
+    '''A class the contains information about a morphology.
+    Its role is to abstract away the raw data.
+    '''
+
+    MTYPE_SEPARATOR = ':'
+
+    def __init__(self, name: str, mtype: str, layer: Optional[Union[str, int]] = None,
+                 label: str = None):
+        '''MorphInfo ctor:
+
+        Upon initialization, all repair attributes are set to True.
+
+        Args:
+            name: the morphology name (without the extension)
+            mtype: the full mtype (ie 'L1_DAC:A')
+            layer: the layer (accepts both string and int)
+            label: (optional) a group label to be used to identify multiple groups of morphologies
+                added to the same collection.
+        '''
+        self.name = name
+        self.mtype = mtype
+        if self.MTYPE_SEPARATOR in mtype:
+            self.mtype_no_subtype, self.msubtype = mtype.split(self.MTYPE_SEPARATOR)
```

Comment:
ok, so we will not use the terminology fullmtype at all? I'm good with it!

Reply:
Yes, I was thinking of using the `mtype` as the fullmtype. What do you think ?

Reply:
I like it. I've seen fullmtype in only a few places, such as in BluePyMM. I've never seen anyone needing the submtype of the mtyp_no_subtype at all, so I think it makes sense to stick to mtype and have  the  split as you  did (or make it optional even)